Create filter for pull request tests on Jenkins

matt-kwong opened the pull request:
Creates a script for `run_tests_matrix.py` to call in order to filter out unnecessary tests based on examining file differences between the branch being tested and gRPC:master.

Currently this script only filters out c++, asan, tsan, and msan tests, and filters by checking the beginning and end of changed files, including its directory, for triggers that indicate whether a pull request can skip certain tests.

This still needs work in adding filters for other tests and feedback on how to prevent incorrectly skipping tests.

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,95 @@
+from subprocess import call, check_output
+
+# triggers to skip c++ tests
+run_cpp_starts_with_triggers = ('src/core',
```
Brainstorming:
perhaps we could divide up the repo logically into several "components". Be default, all files are in the default "affects everything" component (test scripts and src/core definitely need to be in the default component).
  
Then you could specify that src/cpp only triggers c++ tests (not core)
src/csharp only triggers C#
src/python only triggers  python

so we could start with "components" named after languages "c++", "csharp", "python"...  and mark the targets in run_tests_matrix.py accordingly. From the diff, you compute a list of affected components and that gives you what to run (and of course run everything if you've touched in the default component)

I think this is in agreement with that @ctiller  has suggested in an e-mail thread.

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,95 @@
+  any begin or end triggers
+  :param file_names: list of changed files generated by _get_changed_files()
+  :param starts_with_triggers: tuple of strings to match with beginning of file names
+  :param ends_with_triggers: tuple of strings to match with end of file names
```
I don't quite see what you need the "ends_with" triggers for. If you dont' need it now, don't include it and only add it when needed.

I'd say let's just split up the repository in logical subtrees as suggested above.

matt-kwong commented:
I think that should address most of your concerns.

matt-kwong commented:
> nit: compared to the base branch the PR is getting merged into (not necessarily grpc:master)

Updated along with any other instance I mentioned grpc:master

> so if none of the diff files are in one of these subdirs, the script is going to skip c++ tests? That's actually not good enough - if someone decides to change Makefile, that can totally break C++, but this script doesn't capture that (and it's going to be hard to capture this in general). Perhaps we need a different approach.

I changed it to like how you described it in your brainstorm. Right now, unless a changed file matches a whitelist condition, all tests will be run - more whitelist conditions will need to be added and the current ones verified.

I am working on making the whitelist conditions around sanitizer tests to be as fine-grained as possible. Currently, if a change is made only in test/core, only core tests are run. If a change is made in only src/cpp or test/cpp, then all c++ tests AND core sanitizer tests are run. If a src/core change is made, all tests are run. I am not if this is correct at all, so please correct me if I'm wrong. I'd also like to change it so core sanitizer tests are skipped if only test/cpp files are changed.

> dangerous to hardcode origin/master (people might as well merge into v1.0.x or any other branch in which case this gives them wrong results).

This is changed. When we configure Jenkins jobs, the arguments `--filter_pr_tests --base_branch $ghprbTargetBranch` should be added when calling run_tests_matrix.py. 

> string.split() splits on spaces as well and you really should split on newlines only. Also, please check that the splitting works on windows where you might see \r\n line endings (maybe?).

Good catch - I changed it to str.splitlines(), which should work with Windows, but of course I'll test it using Jenkins to make sure.

> I don't quite see what you need the "ends_with" triggers for. If you dont' need it now, don't include it and only add it when needed.

I added a use for this; I think it'll be useful to filter out documentation files such as README.md.

> I'm a bit afraid to skip tests based on a shortname of a test (the shortname is basically a freeform string). Maybe we could come up something more bulletproof. 
On the other hand, perhaps fine for initial experiments (before actually enabling this) if a TODO is added.

TODO added - will address in the future.

matt-kwong commented:
CC @nicolasnoble 

Feedback would be greatly appreciated.

I was thinking about make whitelisting more intuitive by making it so if I wanted to whitelist src/python from non-Python languages, I would add it to the C++, Node, Ruby, Core, PHP, C#, and Objective-C whitelists but this would lead to the file being very long since whitelisting anything for all but one language will require it being added to the file 7 times. Thoughts?

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,253 @@
+
+# Whitelist for all tests
+# If whitelist item should only trigger some tests, the item should be
+# added to this list and the trigger list of tests that should be run
```
I don't like the fact that if you add something to the list here, the tests won't run unless you add it to a list in a different place.

perhaps 
```
{ 
    'test/core': [ CORE_TESTS, ...],
    'src/cpp': [ CPP_TESTS, .... ],
    'summerofcode': [ ],
    'doc': [],
    
}
``` 
is more comprehensible and it is obvious which items don't trigger any tests and which ones do.

jtattermusch replied:
CPP_TESTS  and CORE_TESTS can be objects representing subsets of test to run.

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,253 @@
+                         'test/distrib/python',
+                         'test/distrib/ruby']
+				   
+ends_with_whitelist = ['README.md',
```
nit: would it make sense to use regular expressions to match the path instead of starts_with and ends_with.

`'^src/cpp'` and `'README.md$'` seem quite self-explanatory.

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,253 @@
+  :return: safe to skip tests
+  """
+  # convert lists to tuple to pass into str.startswith() and str.endswith()
+  starts_with_whitelist = tuple(starts_with_whitelist)
```
btw, python has `any` and `all` operators. Using them might result in more readable code.

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,253 @@
+    tests = _remove_irrelevant_tests(tests, '_c_')
+
+  # Filter c++ tests
+  skip_cpp = _can_skip_tests(changed_files,
```
the skip_X part is very repetitive and at some point, wrong triggers will be used for a wrong language.

```
skip_cpp = _can_skip_tests(  /* one day you'll see this */ core_starts_with_triggers)
```

I'd suggest defining and object/class for each test category (core, cpp, ...) and  then 

```
for  testsuite in  [CPP_TESTS, CORE_TESTS, ....]:
  can_skip = can_skip_tests( ....)
  tests = remove_irrelevant_tests( )

````

jtattermusch commented on tools/run_tests/run_tests_matrix.py:
```diff
@@ -262,6 +272,18 @@ def _allowed_labels():
+    print '(TESTING) No tests will be skipped.'
+  else:
+    print '(TESTING) These tests will be skipped:'
+    for job in list(set(jobs) - set(relevant_jobs)):
```
as a TODO we could add skipped tests to the report.xml, because then it's going to be obvious from Jenkins UI what has been skipped.

matt-kwong commented:
I went with all of your suggestions in this commit. Just a few things:
- Changing something like 'src/node/README.md' now triggers Node tests, and this is something I'd like to change without adding too much complexity.
- All of the sanitizer tests will now run or skip; I will change this to be able to specify C vs C++ after reaffirming which situations allow one to be skipped. 
- I will redesign the TestSuite class after deciding how to better identify when a job belongs to a test suite - finding tags in job.shortname is just temporary.

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,184 @@
+    self.triggers.append(trigger)
+
+# Create test suites
+_core_test_suite = TestSuite(['_c_'])
```
nit: I think global constants in python are supposed to be UPPERCASE (like _WHITELIST_DICT). Fine to fix later.

jtattermusch commented on tools/run_tests/filter_pull_request_tests.py:
```diff
@@ -0,0 +1,184 @@
+  '^summerofcode/.*': [],
+  '.*README.md$': [],
+  '.*LICENSE$': [],
+  '^src/cpp.*': [_cpp_test_suite],
```
nit: as above, you probably want src/cpp/.* rather than just src/cpp.*  Also below.

jtattermusch replied:
nit2:  if you are matching just a substring you don't need `.*`   If you are matching the entire filename, you don't need ^ or $.
(I think matching a substring is a bit better)

matt-kwong commented:
Currently this is not working when using Jenkins - I tested it on Windows and Linux. I'll commit changes tomorrow to fix this.
